README.md

# rn-bugly

[![npm version](http://img.shields.io/npm/v/rn-bugly.svg?style=flat-square)](https://npmjs.org/package/rn-bugly "View this project on npm")
[![npm version](http://img.shields.io/npm/dm/rn-bugly.svg?style=flat-square)](https://npmjs.org/package/rn-bugly "View this project on npm")

bugly for react-native,支持统计，android支持应用全量升级

## 安装

> `$ npm install rn-bugly --save`

* react-native <0.60

> `$ react-native link rn-bugly`

* react-native >=0.60

新版RN会自动link,不需要执行link命令

---
不管是哪个版本的react-native，ios端都需要在`ios`文件夹下执行
```shell
$ cd ios
$ pod install
```


### 配置
#### iOS
`AppDelegate.m`
```
#import "RNBugly.h"
-(BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  ...
  //初始化bugly，会自动读取info.plist中的参数
  [RNBugly startWithAppId];
  return YES;
}
```
`info.plist`文件读取SDK初始化参数，可配置的参数如下(`除Appid为必填外，其它可选`)：
```
- Appid
    - Key: BuglyAppIDString
    - Value: 字符串类型
- 渠道标识
    - Key: BuglyAppChannelString
    - Value: 字符串类型
- 版本信息
    - Key: BuglyAppVersionString
    - Value: 字符串类型
- 开启Debug信息显示
    - Key: BuglyDebugEnable
    - Value: BOOL类型
```

#### Android
`MainApplication.java`(依旧可以用Bugly.init进行初始化，下面两个方法只是对Bugly.init的简单封装)
```
import com.reactlibrary.bugly.RNBuglyModule;

@Override
public void onCreate() {
  super.onCreate();

  ...

  ** 下面的方法根据情况二选一 **

  //参数解析：
  //参数1：上下文对象
  //参数2：注册时申请的APPID
  //参数3：是否开启debug模式，true表示打开debug模式，false表示关闭调试模式

  //初始化并且自动检查更新
  RNBuglyModule.init(getApplicationContext(),"注册时申请的APPID",false);

  //仅仅初始化(推荐使用该方法，所有的检查更新触发都在js端，更加灵活)
  RNBuglyModule.initWithoutAutoCheckUpgrade(getApplicationContext(),"注册时申请的APPID",false);
}
```

## 使用
```javascript
import RNBugly from 'rn-bugly';


//如果上面采用的initWithoutAutoCheckUpgrade初始化，并且需要进入app的时候也自动检查更新，可以使用下面的方法手动检查更新
//官方原生的init初始化其实有3s的延迟时间，等待初始化完成，所以js端也需要等待3m后才能检查更新
setTimeout(()=>{
  RNBugly.checkUpgrade({
        isManual: false,
        isSilence: false
      });
},3000);



//点击按钮触发检查更新(一般在[关于]-[检查更新])
//此时早已初始化了，不需要延时
RNBugly.checkUpgrade({
        //该值为true的时候，会有toast提示，如果不需要提示，设置为false
        isManual: true,
        isSilence: false
      });


//获取更新信息
//注意:更新信息获取一次后会储存在本地，即使后台控制停止该版本更新，该方法依旧可以获取到数据
let updateInfo = await RNBugly.getUpgradeInfo();


//上报自定义异常信息
RNBugly.postException({
    errorMsg: 'login error',
    stack: 'sadasdasd\nzxzxsds',
    extraInfo: {
        userInfo: 'test',
        password: '12345'
    }
})
```

bugly后端异常信息如下：

![](https://tva1.sinaimg.cn/large/008eGmZEgy1gpdin9v721j30ww04imxk.jpg)
<br/>
<br/>
![](https://tva1.sinaimg.cn/large/008eGmZEgy1gpdiomwgs7j30w6068aal.jpg)

其他方法及注释请查看: [index.d.ts](types/index.d.ts)

参考[demo](./example),或者下载[example.apk](https://zhaoyang.lanzous.com/ib832sh)体验



## 注意事项&&疑问

#### 1.无法使用?

请升级到最新版本后再试

目前对于React Native不用区分版本，直接使用最新版即可,只是安装方式略有不同

#### 2.<font color='red'>集成后全量更新无效果</font>

请确定targetSDKVersion是否为28或者以上，bugly请求由于使用了http，而android 9默认是不支持http请求的，需要调整下

具体请参考:

https://blog.csdn.net/weixin_34114823/article/details/88037177

在最新的1.1.0的版本中，依赖的官方sdk，已经支持https了，具体查看[官方日志](https://bugly.qq.com/docs/release-notes/release-android-beta/?v=20200622202242)



#### 3.为什么我点击更新按钮后，对话框关闭，啥反应都没有?

等一会会出现安装提示,bugly对的更新方式是直接在通知栏显示下载进度，下载完成覆盖安装，如果状态栏没有提示，那就是没有通知权限(oppo/vivo系统是默认不开启该权限的)
  
## 截图

<img style="width:300px;object-fit: contain;margin-right: 20px;"  src='https://tva1.sinaimg.cn/large/007S8ZIlgy1gdpp54zj0nj30u01uoqed.jpg' />
<img style="width:300px;object-fit: contain" src='https://tva1.sinaimg.cn/large/007S8ZIlgy1gdpp5gjhhvj30u01uotej.jpg' />


CHANGELOG.md
# 1.2.0 (2022-05-24)
* 更新sdk版本号 (@[enix223](https://github.com/enix223))
* 适配RN 0.68+ (@[enix223](https://github.com/enix223))

# 1.1.0 (2021-04-12)
跟0.3.3完全一致，后面在v1进行更新

# 0.3.3 (2021-04-12)
* 固定两端的bugly sdk版本，防止出现版本号不一致导致编译失败问题

# 0.3.2 (2021-04-09)
* 实现ios端的`postException`，修复[#16](https://github.com/yz1311/rn-bugly/issues/16)
* 添加`startCrashReport`、`closeCrashReport`、
  `getCurrentTag`、`getUserData`、`getBuglyVersion`等接口

# 0.3.1 (2021-03-17)
* 由于jcenter仓库的问题，暂时固定`com.tencent.bugly:nativecrashreport`版本为`3.8.0`

jcenter仓库如果无法访问，请使用[aliyun](https://maven.aliyun.com/mvn/guide) 替换

# 0.3.0 (2020-04-11)


## 修复

* 将android端的方法判断，防止ios端调用报错
* 解决./gradlew assemleRelease编译报错


## 新增

* 将初始化方法拆分成两种(`依旧可以用原来的方法进行初始化，下面两个方法只是对Bugly.init的简单封装`)，可以选择进入app时不自动更新，完全由js端代码控制检查
    `原来:`
    ```
    import com.tencent.bugly.Bugly;

    Bugly.init(getApplicationContext(),"注册时申请的APPID",false);
    ```


    `现在:`
    ```
    import com.reactlibrary.bugly.RNBuglyModule;

    ** 下面的方法根据情况二选一 **

    //初始化并且自动检查更新
    RNBuglyModule.init(getApplicationContext(),"注册时申请的APPID",false);

    //仅仅初始化
    RNBuglyModule.initWithoutAutoCheckUpgrade(getApplicationContext(),"注册时申请的APPID",false);
    ```
* 新增example及demo apk，更新文档
* 补全putUserData方法，ios端的实现


example/App.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow strict-local
 */

import React, { useState, useEffect } from 'react';
import {
  SafeAreaView,
  StyleSheet,
  ScrollView,
  View,
  Text,
  StatusBar,
  Button,
  Platform
} from 'react-native';

import {
  Header,
  LearnMoreLinks,
  Colors,
  DebugInstructions,
  ReloadInstructions,
} from 'react-native/Libraries/NewAppScreen';
import RNBugly from 'rn-bugly';

const App: () => React$Node = () => {

  const [updateInfo, setUpdateInfo] = useState(null);

  const checkForUpdate = () => {
    RNBugly.checkUpgrade({
      isManual: false,
      isSilence: false,
    });
  };

  const getUpdateInfo = async () => {
    let updateInfo = await RNBugly.getUpgradeInfo();
    setUpdateInfo(updateInfo);
  };

  useEffect(() => {
    (async () => {
      RNBugly.putUserData('test', 'AAA');
      setTimeout(() => {
        checkForUpdate();
      }, 3000);
    })();
  }, []);

  return (
    <>
      <StatusBar barStyle="dark-content" />
      <SafeAreaView>
        <ScrollView
          contentInsetAdjustmentBehavior="automatic"
          style={styles.scrollView}>
          <Header />
          {global.HermesInternal == null ? null : (
            <View style={styles.engine}>
              <Text style={styles.footer}>Engine: Hermes</Text>
            </View>
          )}
          <View style={styles.body}>
            {Platform.OS == 'android' ?
              <Button title='检查更新' onPress={checkForUpdate}></Button>
              :
              null
            }
            {Platform.OS == 'android' ?
              <View style={{marginTop: 10}}>
                <Button title='获取更新信息' onPress={getUpdateInfo}></Button>
                {updateInfo?
                  <View style={{paddingHorizontal:7,paddingVertical:10}}>
                    {
                      Object.keys(updateInfo).map((key,index)=>(
                        <View key={key} style={{flexDirection:'row'}}>
                            <Text style={{color: '#333',flex:1}}>{({
                              title: '标题',
                              newFeature: '升级说明',
                              publishTime: '发布时间',
                              apkMd5: '安装包Md5',
                              apkUrl: '安装包下载地址',
                              fileSize: '安装包大小',
                              popInterval: '弹窗间隔（ms）',
                              popTimes: '弹窗次数',
                              publishType: '发布类型（0:测试 1:正式）',
                              upgradeType: '弹窗类型（1:建议 2:强制 3:手工）',
                              imageUrl: '图片地址',
                            }[key]||key)+': '}</Text>
                            <Text style={{color: '#999',flex:1}}>{updateInfo[key]}</Text>
                        </View>
                      ))
                    }
                  </View>
                  :
                  null
                }
              </View>
              :
              null
            }
          </View>
        </ScrollView>
      </SafeAreaView>
    </>
  );
};

const styles = StyleSheet.create({
  scrollView: {
    backgroundColor: Colors.lighter,
  },
  engine: {
    position: 'absolute',
    right: 0,
  },
  body: {
    backgroundColor: Colors.white,
  },
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: '600',
    color: Colors.black,
  },
  sectionDescription: {
    marginTop: 8,
    fontSize: 18,
    fontWeight: '400',
    color: Colors.dark,
  },
  highlight: {
    fontWeight: '700',
  },
  footer: {
    color: Colors.dark,
    fontSize: 12,
    fontWeight: '600',
    padding: 4,
    paddingRight: 12,
    textAlign: 'right',
  },
});

export default App;
